File: lib/core/smoothing/ema.dart
class ExponentialMovingAverage {
  ExponentialMovingAverage({required this.alpha})
      : assert(alpha > 0 && alpha <= 1);

  final double alpha;
  double? _value;

  double update(double input) {
    if (_value == null) {
      _value = input;
    } else {
      _value = alpha * input + (1 - alpha) * _value!;
    }
    return _value!;
  }

  double get value => _value ?? 0;

  void reset([double? value]) {
    _value = value;
  }
}

File: src/smoothing/ema.ts
export class ExponentialMovingAverage {
  private value: number | null = null;
  constructor(private alpha: number) {
    if (alpha <= 0 || alpha > 1) {
      throw new Error('alpha must be in (0, 1]');
    }
  }

  update(input: number): number {
    if (this.value === null) {
      this.value = input;
    } else {
      this.value = this.alpha * input + (1 - this.alpha) * this.value;
    }
    return this.value;
  }

  reset(next?: number) {
    this.value = typeof next === 'number' ? next : null;
  }

  get current(): number {
    return this.value ?? 0;
  }
}
